lib/client/mfa/cli.go

```diff
@@ -35,117 +38,155 @@ import (
 	"github.com/gravitational/teleport/lib/auth/webauthnwin"
 )
 
+// CLIMFAType is the CLI display name for an MFA type.
+type CLIMFAType string
+
+const (
+	CLIMFATypeOTP      = "OTP"
+	CLIMFATypeWebauthn = "WEBAUTHN"
+	CLIMFATypeSSO      = "SSO"
+)
+
 // CLIPrompt is the default CLI mfa prompt implementation.
 type CLIPrompt struct {
-	cfg    PromptConfig
-	writer io.Writer
+	PromptConfig
+	// Writer is where the prompt outputs the prompt. Defaults to os.Stderr.
+	Writer io.Writer
+	// AllowStdinHijack allows stdin hijack during MFA prompts.
+	// Stdin hijack provides a better login UX, but it can be difficult to reason
+	// about and is often a source of bugs.
+	// Do not set this options unless you deeply understand what you are doing.
+	// If false then only the strongest auth method is prompted.
+	AllowStdinHijack bool
+	// PreferOTP favors OTP challenges, if applicable.
+	// Takes precedence over AuthenticatorAttachment settings.
+	PreferOTP bool
+	// PreferSSO favors SSO challenges, if applicable.
+	// Takes precedence over AuthenticatorAttachment settings.
+	PreferSSO bool
+	// StdinFunc allows tests to override prompt.Stdin().
+	// If nil prompt.Stdin() is used.
+	StdinFunc func() prompt.StdinReader
 }
 
 // NewCLIPrompt returns a new CLI mfa prompt with the config and writer.
-func NewCLIPrompt(cfg *PromptConfig, writer io.Writer) *CLIPrompt {
+// TODO(Joerger): Delete once /e is no longer dependent on it.
+func NewCLIPrompt(cfg PromptConfig, writer io.Writer) *CLIPrompt {
 	return &CLIPrompt{
-		cfg:    *cfg,
-		writer: writer,
+		PromptConfig: cfg,
+		Writer:       writer,
 	}
 }
 
 func (c *CLIPrompt) stdin() prompt.StdinReader {
-	if c.cfg.StdinFunc == nil {
+	if c.StdinFunc == nil {
 		return prompt.Stdin()
 	}
-	return c.cfg.StdinFunc()
+	return c.StdinFunc()
 }
 
 // Run prompts the user to complete an MFA authentication challenge.
 func (c *CLIPrompt) Run(ctx context.Context, chal *proto.MFAAuthenticateChallenge) (*proto.MFAAuthenticateResponse, error) {
-	if c.cfg.PromptReason != "" {
-		fmt.Fprintln(c.writer, c.cfg.PromptReason)
+	if c.Writer == nil {
+		c.Writer = os.Stderr
 	}
```

Comment:
nit: Could we avoid modifying the instance?

```suggestion
	writer := cmp.Or(c.Writer, os.Stderr)
```

Reply:
(A method like stdin() would do it too.)